dot/state/storage_notify.go

```diff
@@ -33,55 +34,105 @@ type SubscriptionResult struct {
 	}
-
-	var id byte
-	for {
-		id = generateID()
-		if s.subscriptions[id] == nil {
-			break
+	go func() {
+		if err := s.notifyObserver(sr, o); err != nil {
+			logger.Warn("failed to notify storage subscriptions", "error", err)
 		}
-	}
+	}()
 
-	s.changedLock.RUnlock()
-
-	s.changedLock.Lock()
-	s.subscriptions[id] = &sub
-	s.changedLock.Unlock()
-	return id, nil
 }
 
-// UnregisterStorageChangeChannel removes the storage change notification channel with the given ID.
-// A channel must be unregistered before closing it.
-func (s *StorageState) UnregisterStorageChangeChannel(id byte) {
-	s.changedLock.Lock()
-	defer s.changedLock.Unlock()
+// UnregisterStorageObserver removes observer from notification list
+func (s *StorageState) UnregisterStorageObserver(o Observer) {
+	s.observerList = s.removeFromSlice(s.observerList, o)
+}
 
-	delete(s.subscriptions, id)
+func (s *StorageState) notifyAll(root common.Hash) {
+	s.changedLock.RLock()
+	defer s.changedLock.RUnlock()
+	for _, observer := range s.observerList {
+		err := s.notifyObserver(root, observer)
+		if err != nil {
+			logger.Warn("failed to notify storage subscriptions", "error", err)
+		}
+	}
 }
 
-func (s *StorageState) notifyChanged(change *SubscriptionResult) {
-	if len(s.subscriptions) == 0 {
-		return
+func (s *StorageState) notifyObserver(root common.Hash, o Observer) error {
+	s.lock.RLock()
+	t := s.tries[root]
+	defer s.lock.RUnlock()
+
+	if t == nil {
+		return errTrieDoesNotExist(root)
 	}
 
-	logger.Trace("notifying changed storage chans...", "chans", s.subscriptions)
+	subRes := &SubscriptionResult{
+		Hash: root,
+	}
+	if len(o.GetFilter()) == 0 {
+		// no filter, so send all changes
+		ent := t.Entries()
+		for k, v := range ent {
+			if k != ":code" {
+				// todo, currently we're ignoring :code since this is a lot of data
+				kv := &KeyValue{
+					Key:   common.MustHexToBytes(fmt.Sprintf("0x%x", k)),
+					Value: v,
+				}
+				subRes.Changes = append(subRes.Changes, *kv)
+			}
+		}
+	} else {
+		// filter result to include only interested keys
+		for k, cachedValue := range o.GetFilter() {
+			value := t.Get(common.MustHexToBytes(k))
+			if !reflect.DeepEqual(cachedValue, value) {
+				kv := &KeyValue{
+					Key:   common.MustHexToBytes(k),
+					Value: value,
+				}
+				subRes.Changes = append(subRes.Changes, *kv)
+				o.GetFilter()[k] = value
+			}
+		}
+	}
+
+	if len(subRes.Changes) > 0 {
+		logger.Trace("update observer", "changes", subRes.Changes)
+		go func() {
+			o.Update(subRes)
+		}()
+	}
 
-	for _, ch := range s.subscriptions {
-		go func(ch chan<- *SubscriptionResult) {
-			ch <- change
-		}(ch.Listener)
+	return nil
+}
+
+func (s *StorageState) removeFromSlice(observerList []Observer, observerToRemove Observer) []Observer {
+	s.changedLock.Lock()
+	defer s.changedLock.Unlock()
+	observerListLength := len(observerList)
+	for i, observer := range observerList {
+		if observerToRemove.GetID() == observer.GetID() {
```

Comment:
are unique IDs guaranteed?

Reply:
I using a counter that is incremented each time an observer is created.